compiler/rustc_macros/src/diagnostics/diagnostic.rs

```diff
@@ -71,55 +71,42 @@ impl<'a> SessionDiagnosticDerive<'a> {
                     }
                 };
 
+                // Keep track of which fields are subdiagnostics or have no attributes.
+                let mut subdiagnostics_or_empty = std::collections::HashSet::new();
```

Comment:
I'm torn on whether or not we should build up a set like this or just do the same check again but negate the result (putting it into a function to avoid duplication obviously). Any thoughts?

Reply:
I'm under the impression that having the set should be faster than doing the check again but I don't have any hard evidence on it either. I'm happy with both alternatives anyway.

I checked `synstructure` to see if we could generate both at the same time to avoid the second filter but it seems that's not the case.